### PR TITLE
Serenity engine: fall back to eth1 engine before the Merge

### PR DESCRIPTION
**Pull request by yperbasis**
Make Serenity engine suitable for blocks not only after the Merge, but also before.

### CODE REVIEW

**Comment by Giulio2002** [comment]
I do not think we are going to use the serenity consensus engine at all unlike geth, if we backward sync to check block validity we can just check `currentBlock.hash == nextBlock.parenthash`, furthermore block rewards(not tx fees) are now given in the consensus layer to validators rather than in the execution layer , so no block finality on our end. the approach you suggest is much more flexible so if we ever have to use this component i would personally prefer this.

File: consensus/aura/aura.go [hunk 1]
```diff
@@ -697,13 +697,6 @@ func (c *AuRa) verifyFamily(chain consensus.ChainHeaderReader, e consensus.Epoch
 	return nil
 }
 
-// VerifyHeaders is similar to VerifyHeader, but verifies a batch of headers. The
-// method returns a quit channel to abort the operations and a results channel to
-// retrieve the async verifications (the order is that of the input slice).
-func (c *AuRa) VerifyHeaders(chain consensus.ChainHeaderReader, headers []*types.Header, _ []bool) error {
```
Comment (Giulio2002) [inline consensus/aura/aura.go:703]: why did you remove `verifyHeaders`?

File: consensus/serenity/serenity.go [hunk 2]
```diff
@@ -32,31 +29,50 @@ var (
...
+		return s.eth1Engine.Author(header)
+	}
 	return header.Coinbase, nil
 }
 
 // VerifyHeader checks whether a header conforms to the consensus rules of the
 // stock Ethereum serenity engine.
 func (s *Serenity) VerifyHeader(chain consensus.ChainHeaderReader, header *types.Header, seal bool) error {
-	// Retrieve parent from the chain
+	reached, err := IsTTDReached(chain, header.ParentHash, header.Number.Uint64()-1)
```
Comment (Giulio2002) [inline consensus/serenity/serenity.go:68]: POSHeaders ancestor wont have total difficulty unless they are preceeded by the terminal block

File: consensus/serenity/serenity.go [hunk 3]
```diff
@@ -87,54 +95,62 @@ func (s *Serenity) VerifyUncles(chain consensus.ChainReader, header *types.Heade
...
-		enc = append(enc, header.BaseFee)
-	}
-	rlp.Encode(hasher, enc)
-	hasher.Sum(hash[:0])
-	return hash
+	return s.eth1Engine.SealHash(header)
 }
 
 func (s *Serenity) CalcDifficulty(chain consensus.ChainHeaderReader, time, parentTime uint64, parentDifficulty *big.Int, parentNumber uint64, parentHash, parentUncleHash common.Hash, parentSeal []rlp.RawValue) *big.Int {
+	reached, err := IsTTDReached(chain, parentHash, parentNumber)
```
Comment (Giulio2002) [inline consensus/serenity/serenity.go:134]: Again, we need to make sure that if we got a POSHeader before, we do not throw an error

File: consensus/aura/aura.go (hunk 1 above)
Comment (AlexeyAkhunov) [reply consensus/aura/aura.go:703]: Because we are not using it, we only using functions that verify a single header at a time

**Review by AlexeyAkhunov — APPROVED** [review]

File: consensus/serenity/serenity.go (hunk 2 above)
Comment (yperbasis) [reply consensus/serenity/serenity.go:68]: Let's then save total difficulty for all blocks, including PoS ones.

**Comment by yperbasis** [comment]
> I do not think we are going to use the serenity consensus engine at all unlike geth, if we backward sync to check block validity we can just check `currentBlock.hash == nextBlock.parenthash`, furthermore block rewards(not tx fees) are now given in the consensus layer to validators rather than in the execution layer , so no block finality on our end. the approach you suggest is much more flexible so if we ever have to use this component i would personally prefer this.

Consider the execution stage. It calls `engine.Finalize` for block rewards. So if we want the execution stage to function throughout the Merge, we need to supply it with a consensus engine that can transition through the Merge as well (in this particular case, switch from non-zero block rewards to no block rewards).

File: consensus/serenity/serenity.go (hunk 3 above)
Comment (yperbasis) [reply consensus/serenity/serenity.go:134]: I'm going to make a change in a separate PR to write total difficulty for all blocks, incl. PoS ones. I don't want to do it in this PR in order to keep it small.
